### ISSUE TITLE
[addon-knobs] Seeing knobs from previously selected story

### ISSUE THREAD

**Issue by thany** [storybookjs/addon-knobs#19]
**Describe the bug**
When I select the first story after starting storybook, I see its knobs just fine. Then when I select another story, its knobs are *added* to the knobs that were there from the previously selected knobs.

Worse even, if the previously selected story had knobs in categories (tabs), the non-categorized knobs from the newly selected story are now tucked away in a "Other" tab.

**To Reproduce**
Upgrading/migrating from storybook 5.3 to 6.0, I suppose. That's the only thing that happened that made this break. I am already using the knobs preset in `main.js`, rather than its "register" notation in `addons.js`. And of course, everything is at the same 6.0.28 version.

**Expected behavior**
Knobs panel should only ever display knobs from the currently selected story.

**Screenshots**
![image](https://user-images.githubusercontent.com/152227/98098288-6eeac100-1e8e-11eb-8046-5d501074c4e9.png)
1 - These are from the previously selected story and do not belong here. They *never* go away, even after having selected a docs-only story where the side panel goes away entirely.
2 - This is from the currently selected story, and should not be a tab. The currently selected does not have its knobs in categories, so I expect no tabs at all.

**Code snippets**
Relevant part of `main.js`:
```js
module.exports = {
  stories: ['../../src/**/*.stories.@(jsx|mdx)'],
  addons: [
    '@storybook/addon-options',
    '@storybook/addon-knobs',
    '@storybook/addon-a11y',
    '@storybook/addon-docs',
  ]
};
```
Relevant part of `manager.js`:
```js
import { addons } from '@storybook/addons';

// Option defaults:
addons.setConfig({
  name: 'Mijn Enexis',
  showNav: true,
  showPanel: true,
  panelPosition: 'right',
});
```
I don't know what else. There's no reference to knobs in `preview.jsx`, nor in `theme.js`.

**System**
Results of `npx sb@next info`:
```
  System:
    OS: Windows 10 10.0.19041
    CPU: (12) x64 Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
  Binaries:
    Node: 12.19.0 - C:\Program Files\nodejs\node.EXE
    npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 86.0.4240.183
    Edge: Spartan (44.19041.423.0)
  npmPackages:
    @storybook/addon-a11y: 6.0.28 => 6.0.28
    @storybook/addon-actions: 6.0.28 => 6.0.28
    @storybook/addon-docs: 6.0.28 => 6.0.28
    @storybook/addon-knobs: 6.0.28 => 6.0.28
    @storybook/react: 6.0.28 => 6.0.28
    @storybook/theming: 6.0.28 => 6.0.28
```
It doesn't see my actual browser, Firefox 82.

**Comment by thany** [comment]
The workaround (to specifiy `@storybook/addon-knobs/register` instead of `@storybook/addon-knobs`) no longer works. And of course the problem is back. I thought it worked, but it doesn't.

On top of that, changing the value in a knob has no effect on the story that uses it.

@shilman Would you happen to know another trick to get knobs to work again?

**Comment by rohanrp** [comment]
As a temporary fix, you can run some code in an interval to reset the knobs every time a new story has changed. In the below code, all I am doing is waiting for the URL query string 'path' to change and once it does, it will reset the knobs to its original state. This seems to fix the problem.

in your main.js ensure that knobs is your first add-on tab
```javascript
addons: [
    '@storybook/addon-knobs',
      ...
]
```
In your preview.js, add this code to the bottom of the file
```javascript
let currentPath
if (window.parent) {
  const parentWindow = window.parent
  parentWindow.setInterval(function(){
    const urlParams = new URLSearchParams(parentWindow.location.search);
    const path = urlParams.get('path');
    if (path && path !== currentPath) {
      currentPath = path
      
      const knobButtons = parentWindow.document.querySelectorAll('#panel-tab-content button')
      if (knobButtons) {
        const resetButton = knobButtons[knobButtons.length - 1]
        resetButton.click()
      }

    }
  }, 100);
}
```

**Comment by luis-piedrahita** [comment]
> As a temporary fix, you can run some code in an interval to reset the knobs every time a new story has changed. In the below code, all I am doing is waiting for the URL query string 'path' to change and once it does, it will reset the knobs to its original state. This seems to fix the problem.
> 
> in your main.js ensure that knobs is your first add-on tab
> 
> ```js
> addons: [
>     '@storybook/addon-knobs',
>       ...
> ]
> ```
> 
> In your preview.js, add this code to the bottom of the file
> 
> ```js
> let currentPath
> if (window.parent) {
>   const parentWindow = window.parent
>   parentWindow.setInterval(function(){
>     const urlParams = new URLSearchParams(parentWindow.location.search);
>     const path = urlParams.get('path');
>     if (path && path !== currentPath) {
>       currentPath = path
>       
>       const knobButtons = parentWindow.document.querySelectorAll('#panel-tab-content button')
>       if (knobButtons) {
>         const resetButton = knobButtons[knobButtons.length - 1]
>         resetButton.click()
>       }
> 
>     }
>   }, 100);
> }
> ```

Awesome workaround, works very good for me. 
In my case it also has a delay in the number of Knobs displayed in the label, so I added a simple overwrite to the text.

```
const knobLabel = parentWindow.document.querySelector('[id*="tabbutton-knobs-"]',);
knobLabel.textContent = 'Knobs';

**Comment by mgameover** [comment]
Hi everyone, I've created PR with the fix.
In two words, `disconnectCallbacks` is invoked before STORY_CHANGED event has been emitted. As the result the knobs are not being reset.
@shilman , Could you please take a look and merge?

**Comment by luis-piedrahita** [comment]
> Hi everyone, I've created PR with the fix. In two words, `disconnectCallbacks` is invoked before STORY_CHANGED event has been emitted. As the result the knobs are not being reset. @shilman , Could you please take a look and merge?

Works for me the fix!! thanks @mgameover  
I just only have an issue with the amount of knobs displayed when change to a new component, did you realized about it? Tooks the amount of the previous visited component.

**Comment by YaronMiro** [comment]
I also encountered this bug when knobs are not cleared between stories, if using the `@storybook/react` up to version `6.3.13` it works as expected but higher versions are not. At least there is a possibility to upgrade from version 5 to 6 but not the latest version 6.x if intending to use knobs.

**Comment by Shelagh-Lewins** [comment]
I had an issue with knobs for the wrong component (always the same wrong component) being shown on first load, which takes longer to render than switching between components. I modified @rohanrp's solution slightly to allow longer on the first render, and to check for length of the array to avoid console error.

```
let currentPath;
let timeout = 600;

if (window.parent) {
  const parentWindow = window.parent;
  parentWindow.setInterval(function () {
    const urlParams = new URLSearchParams(parentWindow.location.search);
    const path = urlParams.get('path');
    timeout = 150;

    if (path && path !== currentPath) {
      currentPath = path;

      const knobButtons = parentWindow.document.querySelectorAll(
        '#panel-tab-content button',
      );

      if (knobButtons.length) {
        const resetButton = knobButtons[knobButtons.length - 1];
        resetButton.click();
      }
    }
  }, timeout);
}
```

**Comment by literalpie** [comment]
#45 was merged with the fix for this. It is included in the 7.0.2 release
